File: Java/CloneBinaryTree/CloneBinaryTree.cpp
#include<iostream>
#include<unordered_map>
#include<stack>

using namespace std;

class node{
	public:
	int value ;
	node * l ;
	node * r ;
	node * random ;

	node(int val)
	{
		value = val;
		l = r = random = 0;
	}
};

int cloneBinaryTree(node* inputBT, node*& cloneBT ){
	
	if(0 == inputBT) return 0 ;
	cloneBT = new node(inputBT->value) ;

	unordered_map<node*, node*> nodeMapping ;

	stack<node*> stInput ;
	stack<node*> stClone ;
	stInput.push(inputBT) ;
	stClone.push(cloneBT) ;

	// create the clone tree
	while(!stInput.empty())
	{
		node * nI = stInput.top() ;
		stInput.pop() ;

		node * nC = stClone.top() ;
		stClone.pop() ;

		nodeMapping.insert(make_pair(nI, nC)) ;

		if(nI->l){
			stInput.push(nI->l) ;
			
			node* lnode = new node(nI->l->value) ;
			nC->l = lnode ;
			stClone.push(lnode) ;
		}

		if(nI->r){
			stInput.push(nI->r) ;

			node* rnode = new node(nI->r->value) ;
			nC->r = rnode ;
			stClone.push(rnode) ;
		}
	}

	// for random ptr
	for(unordered_map<node*, node*>::iterator it = nodeMapping.begin(); it != nodeMapping.end(); ++ it)
	{
		if(it->first->random)
			it->second->random = nodeMapping.at(it->first->random);
	}

	return 0 ;
}

void printInorder(node* n)
{
	if (n == NULL)
		return;
	     
	/* First recur on left sutree */
	printInorder(n->l);
		 
	/* then print data of Node and its random */
	cout << "[" << n->value << " ";
	if (n->random == NULL)
		cout << "NULL], ";
	else
		cout << n->random->value << "], ";
			     
	/* now recur on right subtree */
	printInorder(n->r);
}

int main(){

	node *tree = new node(1);
	tree->l = new node(2);
	tree->r = new node(3);
	tree->l->l = new node(4);
	tree->l->r = new node(5);
	tree->random = tree->l->r;
	tree->l->l->random = tree;
	tree->l->r->random = tree->r;

	cout << "Inorder traversal of original binary tree is: \n";
	printInorder(tree);
	     
	node *clone = 0 ;
	cloneBinaryTree(tree, clone);
	
	cout << "\n\nInorder traversal of cloned binary tree is: \n";
	printInorder(clone);

	return 0 ;
}

File: Java/CloneBinaryTree/CloneBinaryTree.java
import java.util.*;
class Node
{
	int val;
	Node left, right, random;

	public Node(int val)
	{
		this.val = val;
	}
}
public class CloneBinaryTree
{
	public static Node constructNoRandom(Node root, HashMap<Node,Node> h)
	{
		if(root == null)
			return null;
		Node r = new Node(root.val);
		h.put(r,root.random);

		r.left = constructNoRandom(root.left, h);
		r.right = constructNoRandom(root.right,h);

		return r;

	}

	public static void constructComplete(Node root, HashMap<Node, Node> h)
	{
		if(root != null){
			constructComplete(root.left, h);
			root.random = h.get(root);
			constructComplete(root.right, h);
		}
	}

	public static Node cloneBin(Node root)
	{
		HashMap<Node, Node> map = new HashMap<Node, Node>();
		Node r = constructNoRandom(root, map);

		constructComplete(r,map);

		return r;
	}

	public static void print(Node root)
	{
		if(root != null){

			print(root.left);
			System.out.print(root.val +" ");

			if(root.random != null)
				System.out.print(root.random.val+"\n");
			else
				System.out.print("\n");

			print(root.right);
		}
	}
	public static void main(String args[])
	{
		Node tree = new Node(1);
    	tree.left = new Node(2);
    	tree.right = new Node(3);
    	tree.left.left = new Node(4);
    	tree.left.right = new Node(5);
    	tree.random = tree.left.right;
   		 tree.left.left.random = tree;
    	tree.left.right.random = tree.right;

    	Node r = cloneBin(tree);
    	print(r);
	}

}